In [2]:
#--------------------------------------------------
# Imports
#--------------------------------------------------
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt
import yfinance as yf


In [3]:
#-----------------------------------------------------------------------------
# Calculating the number of trading days in each month of 2020 for AAPL stock
#-----------------------------------------------------------------------------

jan_data = yf.download("AAPL", start="2020-01-01", end="2020-02-01") # downloading the data from the yfinance module for January 2020
jan_trading_days = jan_data.shape[0] # storing the number of rows for January 2020

feb_data = yf.download("AAPL", start="2020-02-01", end="2020-03-01")
feb_trading_days = feb_data.shape[0]

mar_data = yf.download("AAPL", start="2020-03-01", end="2020-04-01")
mar_trading_days = mar_data.shape[0]

apr_data = yf.download("AAPL", start="2020-04-01", end="2020-05-01")
apr_trading_days = apr_data.shape[0]

may_data = yf.download("AAPL", start="2020-05-01", end="2020-06-01")
may_trading_days = may_data.shape[0]

jun_data = yf.download("AAPL", start="2020-06-01", end="2020-07-01")
jun_trading_days = jun_data.shape[0]

jul_data = yf.download("AAPL", start="2020-07-01", end="2020-08-01")
jul_trading_days = jul_data.shape[0]

aug_data = yf.download("AAPL", start="2020-08-01", end="2020-09-01")
aug_trading_days = aug_data.shape[0]    

sep_data = yf.download("AAPL", start="2020-09-01", end="2020-10-01")
sep_trading_days = sep_data.shape[0]

oct_data = yf.download("AAPL", start="2020-10-01", end="2020-11-01")
oct_trading_days = oct_data.shape[0]

nov_data = yf.download("AAPL", start="2020-11-01", end="2020-12-01")
nov_trading_days = nov_data.shape[0]

dec_data = yf.download("AAPL", start="2020-12-01", end="2021-01-01")
dec_trading_days = dec_data.shape[0]

trading_days_2020 = pd.DataFrame({"Month": ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"],
                                   "Trading Days": [jan_trading_days, feb_trading_days, mar_trading_days, apr_trading_days, may_trading_days, jun_trading_days, jul_trading_days, aug_trading_days, sep_trading_days, oct_trading_days, nov_trading_days, dec_trading_days]})
print(trading_days_2020)

total_trading_days = trading_days_2020["Trading Days"].sum()
print(f"Total trading days in 2020: {total_trading_days}")

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed

        Month  Trading Days
0     January            21
1    February            19
2       March            22
3       April            21
4         May            20
5        June            22
6        July            22
7      August            21
8   September            21
9     October            22
10   November            20
11   December            22
Total trading days in 2020: 253


### Number of Trading Days

The above code calculates a total of 253 trading days in 2020, which is less than the 365 calendar days because the stock markets are only open on weekdays and also close on public hoidays.

In [10]:
#-----------------------------------------------------------------------------
# Downloading Apple and Microsoft Data 
# Checking for missing values
#-----------------------------------------------------------------------------

prices = yf.download( # Download historical Apple and Microsoft stock prices
    ["AAPL", "MSFT"],
    start="2020-01-01",
    end="2025-01-01")

print(prices.isna().sum()) # Sum up any missing values


[*********************100%***********************]  2 of 2 completed

Price   Ticker
Close   AAPL      0
        MSFT      0
High    AAPL      0
        MSFT      0
Low     AAPL      0
        MSFT      0
Open    AAPL      0
        MSFT      0
Volume  AAPL      0
        MSFT      0
dtype: int64


### Checking for Missing Values

Historical data from Apple and Microsoft has been downloaded, and the sum of any missing values was calculated to be zero. Therefore, the data from the chosen time period are complete and no further data cleaning is required before analysis.

In [11]:
#-----------------------------
# Basic statistics
#-----------------------------

print(prices.info())

print(prices.describe())

<class 'pandas.DataFrame'>
DatetimeIndex: 1258 entries, 2020-01-02 to 2024-12-31
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   (Close, AAPL)   1258 non-null   float64
 1   (Close, MSFT)   1258 non-null   float64
 2   (High, AAPL)    1258 non-null   float64
 3   (High, MSFT)    1258 non-null   float64
 4   (Low, AAPL)     1258 non-null   float64
 5   (Low, MSFT)     1258 non-null   float64
 6   (Open, AAPL)    1258 non-null   float64
 7   (Open, MSFT)    1258 non-null   float64
 8   (Volume, AAPL)  1258 non-null   int64  
 9   (Volume, MSFT)  1258 non-null   int64  
dtypes: float64(8), int64(2)
memory usage: 108.1 KB
None
Price         Close                      High                       Low  \
Ticker         AAPL         MSFT         AAPL         MSFT         AAPL   
count   1258.000000  1258.000000  1258.000000  1258.000000  1258.000000   
mean     151.377786   285.724092   152.899566   288.482469   149

### Comparing Microsoft and Apple Stock Prices

The above data exploration demonstrates that Microsoft, compared to Apple, had the highest mean stock price at the end of the trading day in over the time period analysed. By considering the High and Low stock values, it can be concluded that there was a greater fluctuation in the Microsoft stock prices throughout the market day on average.

It is important to note that although Microsoft had higher mean stock prices than Apple, this does not necessarily mean that Microsoft had a higher market capitalisation, as this depends on the number of shares available:

Market cap = share price x number of outstanding shares


In [ ]:
#-----------------------------------------------------------------------------
# Calculate daily returns
#-----------------------------------------------------------------------------

daily_returns = prices['Close'].pct_change() * 100  # Calculate daily returns as percentage change

print(daily_returns.head())

print(daily_returns.describe())

Ticker          AAPL      MSFT
Date                          
2020-01-02       NaN       NaN
2020-01-03 -0.972203 -1.245149
2020-01-06  0.796814  0.258487
2020-01-07 -0.470276 -0.911803
2020-01-08  1.608629  1.592854
Ticker         AAPL         MSFT
count   1257.000000  1257.000000
mean       0.118211     0.098782
std        1.995567     1.921144
min      -12.864694   -14.739037
25%       -0.842462    -0.817597
50%        0.118670     0.111111
75%        1.198919     1.094011
max       11.980824    14.216928


### Calculating and Comparing Daily Returns

Above the average daily percentage change in stock prices over the time period has been calculated. This is called the return of a stock, and allows the performance of different stocks to be compared regardless of their starting prices. The values for both companies are similar, but Apple has a higher mean daily return. Apple also has a higher standard deviation of daily returns compared to Microsoft, indicating that the Apple stocks were more volatile.
